feat: add basic table template infrastructure (1 of 3)

seanoliver opened the pull request:
## Summary
Implements foundation for [table quickstart experiment](https://www.notion.so/supabase/Table-Editor-Quickstart-2735004b775f80f680eaeb3c6e581633?source=copy_link) with pre-defined templates.

Users can select from categorized templates to quickly create common table structures.

Resolves GROWTH-504

## Changes
- Add pre-defined table templates for common use cases (social network, e-commerce, blog, to-do list, analytics)
- Create simple template selector UI with category browsing
- Add shared UI components for template selection
- Integrate template selection into TableEditor
- Set up type definitions and utils
- Add localStorage handling for dismiss state
- Add toast notifications for template application

## How to Test
1. Navigate to the **Table Editor** page
2. Click **"Create a table"** button
3. The template selector will appear at the top of the form (if project is less than 7 days old)

### Setting the Feature Flag
To see the template selector, you need to set the feature flag:
1. Go to PostHog (make sure you're in the **staging** project)
2. Find the `tableQuickstart` feature flag
3. Click **Edit**
4. Under **"Optional override"**, select **"templates"**
5. Save the changes

**Note:** In this PR, the quickstart component will ONLY appear when the feature flag is set to `"templates"`. Any other value (including `"control"`, `"ai"`, or undefined) will hide the component.

## Related PRs
- **Part 1 of 3** in the table quickstart feature series
- Part 2: #38934 (View system + AI generation)
- Part 3: #38932 (Telemetry tracking)

## Screenshots
Step 1: Quickstart chooser 
<img width="3114" height="2394" alt="CleanShot 2025-10-10 at 10 50 24@2x" src="https://github.com/user-attachments/assets/321a08c5-370f-48ed-9520-cfd14936d430" />

Step 2: Table selector and autofill
<img width="3312" height="2998" alt="CleanShot 2025-10-10 at 10 52 30@2x" src="https://github.com/user-attachments/assets/31b6e7b9-317c-4c8e-b480-00005b2e382a" />

seanoliver committed:
Use useLocalStorage hook for quickstart dismiss state

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/types.ts:
```diff
@@ -0,0 +1,82 @@
+export type PostgresType =
```
actually shocked we don't have this already 😲

seanoliver replied:
Me too!

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          type: 'uuid',
+          nullable: false,
+          isPrimary: true,
+          default: 'gen_random_uuid()',
```
this default doesn't seem to populate in testing 🤔 (the `now()`s work)

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          nullable: false,
+          unique: true,
+          isForeign: true,
+          references: 'auth.users(id)',
```
I'm guessing foreign keys aren't auto-populated?

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isPrimary: true,
+          default: 'gen_random_uuid()',
+        },
+        { name: 'name', type: 'varchar', nullable: false },
```
```suggestion
        { name: 'name', type: 'text', nullable: false },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          default: 'gen_random_uuid()',
+        },
+        { name: 'name', type: 'varchar', nullable: false },
+        { name: 'slug', type: 'varchar', nullable: false, unique: true },
```
```suggestion
        { name: 'slug', type: 'text', nullable: false, unique: true },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+        { name: 'price', type: 'numeric', nullable: false },
+        { name: 'compare_at_price', type: 'numeric', nullable: true },
+        { name: 'cost', type: 'numeric', nullable: true },
+        { name: 'sku', type: 'varchar', nullable: true, unique: true },
```
```suggestion
        { name: 'sku', type: 'text', nullable: true, unique: true },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+        { name: 'compare_at_price', type: 'numeric', nullable: true },
+        { name: 'cost', type: 'numeric', nullable: true },
+        { name: 'sku', type: 'varchar', nullable: true, unique: true },
+        { name: 'barcode', type: 'varchar', nullable: true },
```
```suggestion
        { name: 'barcode', type: 'text', nullable: true },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isPrimary: true,
+          default: 'gen_random_uuid()',
+        },
+        { name: 'order_number', type: 'varchar', nullable: false, unique: true },
```
```suggestion
        { name: 'order_number', type: 'text', nullable: false, unique: true },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isForeign: true,
+          references: 'profiles(id)',
+        },
+        { name: 'status', type: 'varchar', nullable: false, default: "'pending'" },
```
```suggestion
        { name: 'status', type: 'text', nullable: false, default: "'pending'" },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isPrimary: true,
+          default: 'gen_random_uuid()',
+        },
+        { name: 'title', type: 'varchar', nullable: false },
```
```suggestion
        { name: 'title', type: 'text', nullable: false },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          default: 'gen_random_uuid()',
+        },
+        { name: 'title', type: 'varchar', nullable: false },
+        { name: 'slug', type: 'varchar', nullable: false, unique: true },
```
```suggestion
        { name: 'slug', type: 'text', nullable: false, unique: true },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isForeign: true,
+          references: 'profiles(id)',
+        },
+        { name: 'status', type: 'varchar', nullable: false, default: "'draft'" },
```
```suggestion
        { name: 'status', type: 'text', nullable: false, default: "'draft'" },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isPrimary: true,
+          default: 'gen_random_uuid()',
+        },
+        { name: 'name', type: 'varchar', nullable: false },
```
```suggestion
        { name: 'name', type: 'text', nullable: false },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          default: 'gen_random_uuid()',
+        },
+        { name: 'name', type: 'varchar', nullable: false },
+        { name: 'slug', type: 'varchar', nullable: false, unique: true },
```
```suggestion
        { name: 'slug', type: 'text', nullable: false, unique: true },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+        { name: 'name', type: 'varchar', nullable: false },
+        { name: 'slug', type: 'varchar', nullable: false, unique: true },
+        { name: 'description', type: 'text', nullable: true },
+        { name: 'color', type: 'varchar', nullable: true },
```
```suggestion
        { name: 'color', type: 'text', nullable: true },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isForeign: true,
+          references: 'articles(id)',
+        },
+        { name: 'author_name', type: 'varchar', nullable: false },
```
```suggestion
        { name: 'author_name', type: 'text', nullable: false },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          references: 'articles(id)',
+        },
+        { name: 'author_name', type: 'varchar', nullable: false },
+        { name: 'author_email', type: 'varchar', nullable: false },
```
```suggestion
        { name: 'author_email', type: 'text', nullable: false },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isPrimary: true,
+          default: 'gen_random_uuid()',
+        },
+        { name: 'title', type: 'varchar', nullable: false },
```
```suggestion
        { name: 'title', type: 'text', nullable: false },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+        { name: 'title', type: 'varchar', nullable: false },
+        { name: 'description', type: 'text', nullable: true },
+        { name: 'completed', type: 'bool', nullable: false, default: 'false' },
+        { name: 'priority', type: 'varchar', nullable: true, default: "'medium'" },
```
```suggestion
        { name: 'priority', type: 'text', nullable: true, default: "'medium'" },
```

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isPrimary: true,
+          default: 'gen_random_uuid()',
+        },
+        { name: 'name', type: 'varchar', nullable: false },
```
```suggestion
        { name: 'name', type: 'text', nullable: false },
```

charislam reviewed (commented):
Tested toggling the flag on and off and works as described!

Left a bunch of comments, most minor nits.

Since this is an experiment, let's run with it for now, but if we make this a long-term feature I think there are some UX things that need to be addressed:

 - Having to create tables one-by-one is kind annoying, we should make it possible to select multiple and create them all at once
 - Not having foreign keys could cause problems down the road

But that requires a new UI for table creation, so let's see whether this is a useful nudge at all first 💪🏼

seanoliver commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/templates.ts:
```diff
@@ -0,0 +1,416 @@
+          isForeign: true,
+          references: 'auth.users(id)',
+        },
+        { name: 'username', type: 'varchar', nullable: true, unique: true },
```
Makes sense. We don’t have a strong reason to enforce length restrictions here, so using `text` will reduce confusion and keep things more consistent overall.

seanoliver committed:
Standardize varchar to text and remove local storage keys

seanoliver committed:
Remove gen_random_uuid() defaults from templates

seanoliver committed:
feat(studio): flatten table quickstart selector; show for any new table\n\n- Flattened category→template flow into single-step with category chips\n- Removed first-table-only guard so panel appears for all new tables when flag is on

seanoliver commented:
Update: Incorporated feedback for the table quickstart experiment.\n\n- Show the quickstart panel for any new table (removed first-table-only guard).\n- Flattened the category→template flow into a single-step selector with category chips.\n\nKeeping multi-select creation and automatic foreign keys out of scope for now to keep the experiment lean.

seanoliver commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableEditor.tsx:
```diff
@@ -87,6 +92,30 @@ export const TableEditor = ({
   const { mutate: sendEvent } = useSendEventMutation()
+  const tableQuickstartVariant = usePHFlag('tableQuickstart') as
+    | QuickstartVariant
+    | false
```
Ah yeah sorry this is unclear - adding some comments to clarify the difference: usePHFlag returns false when posthog has initialized but the user either isn't targeted yet or the flag itself hasn't been turned on yet; it returns `undefined` when posthog is first initializing and getting the data.

seanoliver committed:
add comments to clarify diff btwn false and undefined in feature flag response type

seanoliver committed:
refactor: improve performance and type safety in table quickstart

- Memoize categories and displayed templates to prevent unnecessary re-renders
- Use generic type parameter instead of type assertion for feature flag
- Simplify dismiss handler with inline function
- Add JSDoc comment explaining feature flag return states

seanoliver committed:
Mark variant param as intentionally unused for future PR

seanoliver committed:
chore: clean up unused import in TableEditor

Remove unused useTablesQuery import after switching to localStorage-based dismissal.

seanoliver committed:
add 7 day project creation gate and fixed uuid default

seanoliver committed:
Improve table template selector UX with success feedback and clearer copy

- Add success toast when template is applied
- Update heading from "Quickstart: choose a template" to "Start faster with a table template"
- Improve subheading to explain time-saving benefit
- Use monospace font for table names to match code convention
- Update template rationales to be more descriptive and specific
- Increase font sizes for better readability
- Remove redundant category display
- Reorder imports alphabetically

seanoliver committed:
Apply prettier formatting to TableTemplateSelector

seanoliver committed:
Merge remote-tracking branch 'origin/master' into feat/table-quickstart-basic-templates

# Conflicts:
#	apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableEditor.tsx
